sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java

```diff
@@ -160,150 +153,135 @@ public void forceFlush() {
+    private long nextExportTime;
+
+    private final BlockingQueue<ReadableSpan> queue;
+
+    private final AtomicReference<CompletableResultCode> flushRequested = new AtomicReference<>();
+    private volatile boolean continueWork = true;
+    private final ArrayList<SpanData> batch;
 
     private Worker(
         SpanExporter spanExporter,
         long scheduleDelayMillis,
-        int maxQueueSize,
         int maxExportBatchSize,
-        int exporterTimeoutMillis) {
+        int exporterTimeoutMillis,
+        BlockingQueue<ReadableSpan> queue) {
       this.spanExporter = spanExporter;
-      this.scheduleDelayMillis = scheduleDelayMillis;
-      this.maxQueueSize = maxQueueSize;
-      this.halfMaxQueueSize = maxQueueSize >> 1;
+      this.scheduleDelayNanos = TimeUnit.MILLISECONDS.toNanos(scheduleDelayMillis);
       this.maxExportBatchSize = maxExportBatchSize;
       this.exporterTimeoutMillis = exporterTimeoutMillis;
-      this.spansList = new ArrayList<>(maxQueueSize);
+      this.queue = queue;
+      this.batch = new ArrayList<>(this.maxExportBatchSize);
     }
 
     private void addSpan(ReadableSpan span) {
-      synchronized (monitor) {
-        if (spansList.size() == maxQueueSize) {
-          droppedSpans.add(1);
-          return;
-        }
-        // TODO: Record a gauge for referenced spans.
-        spansList.add(span);
-        // Notify the worker thread that at half of the queue is available. It will take
-        // time anyway for the thread to wake up.
-        if (spansList.size() >= halfMaxQueueSize) {
-          monitor.notifyAll();
-        }
+      if (!queue.offer(span)) {
+        droppedSpans.add(1);
       }
     }
 
     @Override
     public void run() {
-      while (!Thread.currentThread().isInterrupted()) {
-        // Copy all the batched spans in a separate list to release the monitor lock asap to
-        // avoid blocking the producer thread.
-        ArrayList<ReadableSpan> spansCopy;
-        synchronized (monitor) {
-          // If still maxExportBatchSize elements in the queue better to execute an extra
-          if (spansList.size() < maxExportBatchSize) {
-            do {
-              // In the case of a spurious wakeup we export only if we have at least one span in
-              // the batch. It is acceptable because batching is a best effort mechanism here.
-              try {
-                monitor.wait(scheduleDelayMillis);
-              } catch (InterruptedException ie) {
-                // Preserve the interruption status as per guidance and stop doing any work.
-                Thread.currentThread().interrupt();
-                return;
-              }
-            } while (spansList.isEmpty());
+      updateNextExportTime();
+
+      while (continueWork) {
+        try {
+          if (flushRequested.get() != null) {
+            flush();
+          }
+
+          ReadableSpan lastElement = queue.poll(100, TimeUnit.MILLISECONDS);
```

Comment:
I think it's only this line that can produce a `InterruptedException`, can you narrow down the scope of the try / catch to make that more obvious?